analyzere/base_resources.py

```diff
@@ -263,6 +265,9 @@ def upload_data(self, file_or_str):
             headers = {'Offset': str(offset),
                        'Content-Type': 'application/offset+octet-stream'}
             request_raw('patch', self._data_path, headers=headers, body=chunk)
+            # if there is a known size, and an upload callback, call it
+            if length and upload_callback:
```

Comment:
You might want to validate the upload_callback and commit_callback arguments and raise some kind of error early on if they are not None and not callable().

`callable` is missing (ridiculous) from python [3.0, 3.1], but hopefully we don't care;  there are workarounds if we do.

one possible approach:
```
def upload_data(self, file_or_str, upload_callback=lambda x: None, commit_callback=lambda x: None):
   if not callable(upload_callback):
     raise whatever
   if not callable(commit_callback):
      raise whatever
```
And the subsequent if checks on the functions themselves go away below

Note: you could just replace the default Nones with the lambda and then you only need to validate the callable aspect.  bonus points for validating that the function takes the one argument you expect.